### DISCUSSION TITLE
Accessibility roadmap

### DISCUSSION THREAD

**Discussion by javierm** [consuldemocracy/consuldemocracy#5510] [General]
Currently, the biggest know accessibility issue we've got is the fact that not every user can add annotations in legislation processes (issue #4831).

However, there are other accessibility issues that make it harder to use our site. Even if (on average) we had one or two issues per page, since we've got hundreds of pages, that would mean hundreds of accessibility issues.

There are many automated tools checking for accessibility issues. However, most of them have two important issues:

1. They don't check content that can only be accessed by signed in users
2. They report false positives

So, while these tools might be useful to a certain point, they're pretty limited.

The [axe accessibility testing engine](https://github.com/dequelabs/axe-core) solves both issues, since it can be executed while running our test suite (meaning we can sign in users with any role before running it) and it's built with a "zero false positives" philosophy in mind.

Since right now running axe in our test suite reports too many errors, we could take the following approach.

First, run the test suite taking only critical issues into account, by applying the following patch:

```diff
diff --git a/spec/rails_helper.rb b/spec/rails_helper.rb
index 16a958e386..f0d63bd41e 100644
--- a/spec/rails_helper.rb
+++ b/spec/rails_helper.rb
@@ -21,6 +21,18 @@ require "capybara/rspec"
 require "selenium/webdriver"
 require "view_component/test_helpers"
 
+module Axe
+  module Matchers
+    class BeAxeClean
+      def audit(page)
+        @audit ||= Core.new(page).call(@run).tap do |audit|
+          audit.results.violations.select! { |violation| violation.impact == :critical }
+        end
+      end
+    end
+  end
+end
+
 module ViewComponent
   module TestHelpers
     def sign_in(user)
@@ -60,6 +72,8 @@ module Capybara
     def visit(url, ...)
       original_visit(url, ...)
 
+      expect(page).to be_axe_clean unless driver.name == :rack_test
+
       unless url.match?("robots.txt") || url.match?("active_storage/representations")
         expect(page).to have_css "main", count: 1
         expect(page).to have_css "#main", count: 1
diff --git a/spec/spec_helper.rb b/spec/spec_helper.rb
index a82d425e6a..251fcb340f 100644
--- a/spec/spec_helper.rb
+++ b/spec/spec_helper.rb
@@ -1,3 +1,4 @@
+require "axe-rspec"
 require "factory_bot_rails"
 require "email_spec"
 require "devise"
```

With this, only critical errors are reported. We can have a look at the errors we get and then tackle them one at a time. For example, we could start checking that all form inputs are associated with labels (pull request #5509). That's done by replacing the previous patch with this one:

```diff
diff --git a/spec/rails_helper.rb b/spec/rails_helper.rb
index 16a958e386..d33455f384 100644
--- a/spec/rails_helper.rb
+++ b/spec/rails_helper.rb
@@ -60,6 +60,10 @@ module Capybara
     def visit(url, ...)
       original_visit(url, ...)
 
+      unless driver.name == :rack_test
+        expect(page).to be_axe_clean.checking_only :label
+      end
+
       unless url.match?("robots.txt") || url.match?("active_storage/representations")
         expect(page).to have_css "main", count: 1
         expect(page).to have_css "#main", count: 1
diff --git a/spec/spec_helper.rb b/spec/spec_helper.rb
index a82d425e6a..251fcb340f 100644
--- a/spec/spec_helper.rb
+++ b/spec/spec_helper.rb
@@ -1,3 +1,4 @@
+require "axe-rspec"
 require "factory_bot_rails"
 require "email_spec"
 require "devise"
```

Then, we could solve those issues, run the test suite with the patch that detects critical issues once again, and continue the process until we've solved all critical issues. Then, repeat for the serious issues.

An important thing to take into account is that the fact that the report generated by Axe doesn't contain any false positives means it doesn't report all possible issues. **We also need to keep working on accessibility issues that are not detected by Axe**.

Another important question is: what should we do regarding automated testing once we solve a certain problem? Testing accessibility on every page under every condition is great, but it would make our test suite much slower. Should we only execute these tests before a release, in order to check for regressions? But when should we run the automated accessibility tests on pull requests adding new features?

**Comment by javierm** [comment]
@decabeza @microweb10 Any thoughts?

**Reply by microweb10** [reply]
Hello,

Using Axe seems a good idea. 👌 

I also agree with the approach you mentioned about first checking for critical issues, then serious issues, and so on. 😌 

About when to run the accessibility tests, that's not an easy question to answer. 😅 
From one side I would prefer to run them on every PR, but if that will slow the test suite a lot, maybe we can create a new PR label and only run the accessibility test when the PR contains that label. 🤔
